### PR TITLE
support export after save model

### PR DESCRIPTION
**Pull request by zhangyubo0722**

### CODE REVIEW

File: ppdet/engine/trainer.py [hunk 1]
```diff
@@ -75,7 +75,7 @@ def __init__(self, cfg, mode='train'):
         self.custom_white_list = self.cfg.get('custom_white_list', None)
         self.custom_black_list = self.cfg.get('custom_black_list', None)
         self.use_master_grad = self.cfg.get('master_grad', False)
-        if 'slim' in cfg and cfg['slim_type'] == 'PTQ':
+        if ('slim' in cfg and cfg['slim_type'] == 'PTQ') or 'model_export_enabled' in cfg:
```
Comment (zhangyubo0722) [inline ppdet/engine/trainer.py]: 更改变量名，统一模型输出

File: ppdet/engine/trainer.py [hunk 2]
```diff
@@ -1212,8 +1212,13 @@ def _get_infer_cfg_and_input_spec(self,
                 "img_name": str,
             })
         if prune_input:
+            if self.cfg.get("model_export_enabled", False):
```
Comment (zhangyubo0722) [inline ppdet/engine/trainer.py]: self.xxx.   不做判断

File: ppdet/engine/trainer.py [hunk 3]
```diff
@@ -1490,3 +1495,18 @@ def reset_norm_param_attr(self, layer, **kwargs):
                     setattr(layer, name, new_sublayer)
 
         return layer
+
+class ExportModel(nn.Layer):
+    def __init__(self, model):
+        super().__init__()
+        self.base_model = model
+
+    def eval(self):
```
Comment (zhangyubo0722) [inline ppdet/engine/trainer.py]: 是否必要

File: ppdet/engine/callbacks.py [hunk 4]
```diff
@@ -250,10 +255,12 @@ def on_epoch_end(self, status):
                         save_model(
                             status['weight'],
                             self.model.optimizer,
-                            self.save_dir,
+                            os.path.join(self.save_dir, save_name) if self.model_export_enabled else self.save_dir,
                             save_name,
                             epoch_id + 1,
                             ema_model=weight)
+                        if self.model_export_enabled:
```
Comment (zhangyubo0722) [inline ppdet/engine/callbacks.py]: ema 静态图， else

File: ppdet/engine/export_utils.py [hunk 5]
```diff
@@ -366,7 +366,8 @@ def _dump_infer_config(config, path, image_shape, model):
     else:
         reader_cfg = config['TestReader']
         dataset_cfg = config['TestDataset']
-
+    # print(dataset_cfg)
+    # exit()
```
Comment (zhangyubo0722) [inline ppdet/engine/export_utils.py]: 删掉

File: ppdet/engine/callbacks.py [hunk 6]
```diff
@@ -226,8 +227,15 @@ def on_epoch_end(self, status):
...
-                    save_path = os.path.join(self.save_dir, f"{save_name}.pdstates")
+                    save_path = os.path.join(os.path.join(self.save_dir, save_name) if self.uniform_output_enabled else self.save_dir, f"{save_name}.pdstates")
                     paddle.save(epoch_metric, save_path)
+                    if self.uniform_output_enabled:
+                        save_model_info(epoch_metric, self.save_dir, save_name)
+                        if epoch_id + 1 == self.model.cfg.epoch:
+                            done_flag = True
+                        else:
+                            done_flag = False
+                        update_train_results(self.model.cfg, save_name, epoch_metric, done_flag=done_flag, ema=self.model.use_ema)
```
Comment (TingquanGao) [inline ppdet/engine/callbacks.py]: epoch_id + 1 == self.model.cfg.epoch

File: ppdet/engine/trainer.py [hunk 7]
```diff
@@ -75,13 +76,27 @@ def __init__(self, cfg, mode='train'):
...
             self.log_ranks = [int(i) for i in log_ranks.split(',')]
         elif isinstance(log_ranks, int):
             self.log_ranks = [log_ranks]
+        train_results_path = os.path.abspath(os.path.join(self.cfg.save_dir, "train_results.json"))
+        if self.uniform_output_enabled:
+            if os.path.exists(train_results_path) and self.mode == 'train':
+                try:
+                    os.remove(train_results_path)
+                except:
+                    pass
```
Comment (TingquanGao) [inline ppdet/engine/trainer.py:93]: pass

File: ppdet/engine/trainer.py [hunk 8]
```diff
@@ -1136,9 +1151,12 @@ def _get_infer_cfg_and_input_spec(self,
                                       save_dir,
                                       prune_input=True,
                                       kl_quant=False,
-                                      yaml_name=None):
+                                      yaml_name=None,
+                                      base_model=None):
```
Comment (TingquanGao) [inline ppdet/engine/trainer.py]: base_model -> model

File: ppdet/engine/trainer.py [hunk 9]
```diff
@@ -1251,6 +1268,7 @@ def export(self, output_dir='output_inference', for_fd=False):
         if hasattr(self.model, 'aux_head'):
             self.model.__delattr__('aux_head')
         self.model.eval()
+        base_model = copy.deepcopy(self.model)
```
Comment (TingquanGao) [inline ppdet/engine/trainer.py]: base_model  model

File: ppdet/engine/trainer.py [hunk 10]
```diff
@@ -1490,3 +1508,12 @@ def reset_norm_param_attr(self, layer, **kwargs):
...
         return layer
+
+class ExportModel(nn.Layer):
+    def __init__(self, model):
+        super().__init__()
+        self.base_model = model
+
+    def forward(self, x):
+        x = self.base_model(x)
+        return x
```
Comment (TingquanGao) [inline ppdet/engine/trainer.py]: 去掉

File: ppdet/utils/checkpoint.py [hunk 11]
```diff
@@ -26,6 +27,13 @@
 from .logger import setup_logger
 logger = setup_logger(__name__)
 
+def convert_to_dict(obj):
+    if isinstance(obj, dict):
+        return {k: convert_to_dict(v) for k, v in obj.items()}
+    elif isinstance(obj, list):
```
Comment (TingquanGao) [inline ppdet/utils/checkpoint.py:33]: elif isinstance(obj, list): -> tuple

File: ppdet/utils/checkpoint.py [hunk 12]
```diff
@@ -375,3 +383,69 @@ def save_semi_model(teacher_model, student_model, optimizer, save_dir,
...
+
+def update_train_results(config,
+                         prefix,
+                         metric_info,
+                         done_flag=False,
+                         k=5,
+                         ema=False):
+    if paddle.distributed.get_rank() != 0:
+        return
+    assert k >= 1
```
Comment (TingquanGao) [inline ppdet/utils/checkpoint.py]: last_num

**Review by TingquanGao — APPROVED** [review]

**Pull request merged**